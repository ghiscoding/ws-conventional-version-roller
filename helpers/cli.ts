import execa from 'execa';
import path from 'path';

import lernaCLI from '../packages/cli/src/lerna-cli';

const LERNA_BIN = path.resolve(__dirname, '../packages/cli/src/cli.ts');

/**
 * A higher-order function to help with passing _actual_ yargs-parsed argv
 * into command constructors (instead of artificial direct parameters).
 *
 * @param {Object} commandModule The yargs command exports
 * @return {Function} with partially-applied yargs config
 */
export function commandRunner(commandModule: any) {
  /* eslint-disable import/no-dynamic-require, global-require */
  const cmd = commandModule.command.split(' ')[0];

  // prime the pump so slow-as-molasses CI doesn't fail with delayed require()
  require(path.resolve((require as any).main.filename, '../..'));

  return (cwd: string) => {
    // create a _new_ yargs instance every time cwd changes to avoid singleton pollution
    const cli = lernaCLI([], cwd)
      .exitProcess(false)
      .detectLocale(false)
      .showHelpOnFail(false)
      .wrap(null)
      .command(commandModule);

    return (...args: string[]) =>
      new Promise((resolve, reject) => {
        const yargsMeta: any = {};
        const context = {
          cwd,
          lernaVersion: '__TEST_VERSION__',
          loglevel: 'silent',
          progress: false,
          onResolved: (result) => {
            // success resolves the result, if any, returned from execute()
            resolve(Object.assign({}, result, yargsMeta));
          },
          onRejected: (result) => {
            Object.assign(result, yargsMeta);
            // tests expect errors thrown to indicate failure,
            // _not_ just non-zero exitCode
            reject(result);
          },
        };

        const parseFn = (_yargsError: Error | undefined, parsedArgv: any, yargsOutput: string) => {
          // this is synchronous, before the async handlers resolve
          Object.assign(yargsMeta, { parsedArgv, yargsOutput });
        };

        cli
          .fail((msg: string, err: Error) => {
            // since yargs 10.1.0, this is the only way to catch handler rejection
            // _and_ yargs validation exceptions when using async command handlers
            const actual = (err || new Error(msg)) as Error & { exitCode: number };
            // backfill exitCode for test convenience
            yargsMeta.exitCode = 'exitCode' in actual ? actual.exitCode : 1;
            context.onRejected(actual);
          })
          .parse([cmd, ...args], context, parseFn);
      });
  };
}

exports.cliRunner = function cliRunner(cwd: string, env: { [key: string]: string }) {
  const opts = {
    cwd,
    env: Object.assign(
      {
        CI: 'true',
        // always turn off chalk
        FORCE_COLOR: '0',
      },
      env
    ),
    // when debugging integration test snapshots, uncomment next line
    // stdio: ["ignore", "inherit", "inherit"],
  };

  return (...args: string[]) => execa('node', [LERNA_BIN].concat(args), opts);
};
