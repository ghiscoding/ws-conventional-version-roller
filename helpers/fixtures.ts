import execa from 'execa';
import fileUrl from 'file-url';
import tempy from 'tempy';
import fs from 'fs-extra';
import findUp from 'find-up';
import path from 'path';
import { gitAdd, gitCommit, gitInit } from './git';

export function cloneFixtureFactory(startDir: string) {
  const initFixture = initFixtureFactory(startDir);

  return (...args: any[]) =>
    // @ts-ignore
    initFixture(...args).then((cwd) => {
      const repoDir = tempy.directory();
      const repoUrl = fileUrl(repoDir, { resolve: false });

      return execa('git', ['init', '--bare'], { cwd: repoDir })
        .then(() => execa('git', ['checkout', '-B', 'main'], { cwd }))
        .then(() => execa('git', ['remote', 'add', 'origin', repoUrl], { cwd }))
        .then(() => execa('git', ['push', '-u', 'origin', 'main'], { cwd }))
        .then(() => ({
          cwd,
          repository: repoUrl,
        }));
    });
}

export function findFixture(cwd: string, fixtureName: string) {
  return findUp(path.join('__fixtures__', fixtureName), { cwd, type: 'directory' }).then((fixturePath) => {
    if (fixturePath === undefined) {
      throw new Error(`Could not find fixture with name "${fixtureName}"`);
    }

    return fixturePath;
  });
}

export function copyFixture(targetDir: string, fixtureName: string, cwd: string) {
  return findFixture(cwd, fixtureName).then((fp) => fs.copy(fp, targetDir));
}

export function initFixtureFactory(startDir: string) {
  return (fixtureName: string, commitMessage: boolean | string = 'Init commit') => {
    const cwd = tempy.directory();
    let chain: Promise<any> = Promise.resolve();

    chain = chain.then(() => process.chdir(cwd));
    chain = chain.then(() => copyFixture(cwd, fixtureName, startDir));
    chain = chain.then(() => gitInit(cwd, '.'));

    if (commitMessage) {
      chain = chain.then(() => gitAdd(cwd, '-A'));
      chain = chain.then(() => gitCommit(cwd, commitMessage));
    }

    return chain.then(() => cwd);
  };
}

export function initNamedFixtureFactory(startDir: string) {
  return (dirName: string, fixtureName: string, commitMessage: boolean | string = 'Init commit') => {
    const cwd = path.join(tempy.directory(), dirName);
    let chain: Promise<any> = Promise.resolve();

    chain = chain.then(() => fs.ensureDir(cwd));
    chain = chain.then(() => process.chdir(cwd));
    chain = chain.then(() => copyFixture(cwd, fixtureName, startDir));
    chain = chain.then(() => gitInit(cwd, '.'));

    if (commitMessage) {
      chain = chain.then(() => gitAdd(cwd, '-A'));
      chain = chain.then(() => gitCommit(cwd, commitMessage));
    }

    return chain.then(() => cwd);
  };
}
