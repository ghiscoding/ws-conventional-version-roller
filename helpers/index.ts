import normalizeNewline from 'normalize-newline';
import normalizePath from 'normalize-path';
import { relative } from 'node:path';

import { Project } from '../packages/core/src/project/index.js';

/**
 * Update lerna config inside a test case.
 *
 * @param {String} testDir where target lerna.json exists
 * @param {Object} updates mixed into existing JSON via Object.assign
 */
export function updateLernaConfig(testDir, updates) {
  const project = new Project(testDir);

  Object.assign(project.config, updates);

  return project.serializeConfig();
}

export function multiLineTrimRight(str) {
  return normalizeNewline(str)
    .split('\n')
    .map((line) => line.trimRight())
    .join('\n');
}

export function normalizeRelativeDir(testDir, filePath) {
  return normalizePath(relative(testDir, filePath));
}

export * from './git/index.js';
export * from './fixtures.js';
export * from './npm/index.js';
export * from './cli.js';
export * from './logging-output.js';
export * from './pkg-matchers.js';
