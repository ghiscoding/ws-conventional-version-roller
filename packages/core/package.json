{
  "name": "@lerna-lite/core",
  "description": "Lerna-Lite core implementation module",
  "version": "2.1.0",
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "/dist"
  ],
  "type": "module",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "module": "dist/index.js",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "default": "./dist/index.js",
      "types": "./dist/index.d.ts"
    },
    "./*": "./*",
    "./cli.js": "./dist/cli.js",
    "./dist/cli.js": "./dist/cli.js"
  },
  "typesVersions": {
    "*": {
      "*": [
        "./dist/index.d.ts"
      ]
    }
  },
  "scripts": {
    "prebuild": "rimraf dist",
    "build": "tsc",
    "build:incremental": "tsc --incremental --sourceMap false --excludeDirectories dist",
    "pack-tarball": "npm pack"
  },
  "license": "MIT",
  "author": "Ghislain B.",
  "homepage": "https://github.com/lerna-lite/lerna-lite",
  "repository": {
    "type": "git",
    "url": "https://github.com/lerna-lite/lerna-lite.git",
    "directory": "packages/core"
  },
  "bugs": {
    "url": "https://github.com/lerna-lite/lerna-lite/issues"
  },
  "engines": {
    "node": ">=16.15.0",
    "npm": ">=8.5.0"
  },
  "dependencies": {
    "@npmcli/run-script": "^6.0.0",
    "chalk": "^5.2.0",
    "clone-deep": "^4.0.1",
    "config-chain": "^1.1.13",
    "cosmiconfig": "^8.1.3",
    "dedent": "^0.7.0",
    "execa": "^7.1.1",
    "fs-extra": "^11.1.1",
    "glob-parent": "^6.0.2",
    "globby": "^13.1.4",
    "inquirer": "^9.1.5",
    "is-ci": "^3.0.1",
    "json5": "^2.2.3",
    "load-json-file": "^7.0.1",
    "minimatch": "^9.0.0",
    "npm-package-arg": "^10.1.0",
    "npmlog": "^7.0.1",
    "p-map": "^6.0.0",
    "p-queue": "^7.3.4",
    "resolve-from": "^5.0.0",
    "semver": "^7.5.0",
    "slash": "^5.0.0",
    "strong-log-transformer": "^2.1.0",
    "write-file-atomic": "^5.0.0",
    "write-json-file": "^5.0.0",
    "write-pkg": "^5.1.0"
  },
  "devDependencies": {
    "@types/clone-deep": "^4.0.1",
    "@types/dedent": "^0.7.0",
    "@types/fs-extra": "^11.0.1",
    "@types/glob-parent": "^5.1.1",
    "@types/inquirer": "^9.0.3",
    "@types/minimatch": "^5.1.2",
    "@types/npm-package-arg": "^6.1.1",
    "@types/semver": "^7.3.13",
    "@types/strong-log-transformer": "^1.0.0",
    "@types/write-file-atomic": "^4.0.0",
    "npm-registry-fetch": "^14.0.4",
    "yargs": "^17.7.1",
    "yargs-parser": "^21.1.1"
  },
  "funding": {
    "type": "ko_fi",
    "url": "https://ko-fi.com/ghiscoding"
  }
}
