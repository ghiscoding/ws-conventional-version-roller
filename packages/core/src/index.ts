// folders
export * from './models';
export * from './package-graph';
export * from './project';
export * from './utils';

// files
export * from './child-process';
export * from './constants';
export * from './command';
export * from './package';
export * from './prompt';
export * from './validation-error';
