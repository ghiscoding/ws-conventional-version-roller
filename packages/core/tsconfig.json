{
  "extends": "../../tsconfig.base.json",
  "compileOnSave": false,
  "compilerOptions": {
    "module": "ESNext",
    "rootDir": "src",
    "outDir": "dist",
    "types": ["node"]
  },
  "filesGlob": ["./src/**/*.ts"],
  "include": ["src/**/*.ts"],
  "exclude": ["dist", "node_modules", "**/__helpers__/**", "**/__mocks__/**", "**/__tests__/**"]
}
