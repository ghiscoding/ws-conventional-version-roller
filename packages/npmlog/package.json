{
  "name": "@lerna-lite/npmlog",
  "description": "Lerna-Lite npmlog reimplementation of the now deprecated npm/npmlog",
  "version": "3.5.1",
  "files": [
    "/dist"
  ],
  "type": "module",
  "publishConfig": {
    "access": "public"
  },
  "types": "./dist/index.d.ts",
  "exports": {
    "types": "./dist/index.d.ts",
    "default": "./dist/index.js"
  },
  "scripts": {
    "prebuild": "rimraf dist",
    "build": "tsc",
    "build:incremental": "tsc --incremental --sourceMap false --excludeDirectories dist",
    "pack-tarball": "npm pack"
  },
  "license": "MIT",
  "author": "Ghislain B.",
  "homepage": "https://github.com/lerna-lite/lerna-lite",
  "repository": {
    "type": "git",
    "url": "https://github.com/lerna-lite/lerna-lite.git",
    "directory": "packages/npmlog"
  },
  "bugs": {
    "url": "https://github.com/lerna-lite/lerna-lite/issues"
  },
  "engines": {
    "node": "^18.0.0 || >=20.0.0"
  },
  "dependencies": {
    "are-we-there-yet": "^4.0.2",
    "console-control-strings": "^1.1.0",
    "gauge": "^5.0.2",
    "set-blocking": "^2.0.0"
  },
  "devDependencies": {
    "yargs-parser": "^21.1.1"
  }
}
