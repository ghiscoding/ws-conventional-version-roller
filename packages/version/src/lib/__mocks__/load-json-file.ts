import { afterEach, vi } from 'vitest';
import { dirname } from 'node:path';
import normalizePath from 'normalize-path';

const { loadJsonFile: actualLoadJsonFile, loadJsonFileSync: loadJsonFileSyncActual } =
  await vi.importActual<any>('load-json-file');

const asyncRegistry = new Map();
const syncRegistry = new Map();

function incrementCalled(registry, manifestLocation) {
  // tempy creates dirnames that are 32 characters long, but we want a readable key
  const subPath = (manifestLocation || '').split(/[0-9a-f]{32}/).pop();
  const key = normalizePath(dirname(subPath));

  // keyed off directory subpath, _not_ pkg.name (we don't know it yet)
  registry.set(key, (registry.get(key) || 0) + 1);
}

// by default, act like a spy that counts number of times each location was loaded
export const loadJsonFile: any = vi.fn((manifestLocation) => {
  incrementCalled(asyncRegistry, manifestLocation);

  return actualLoadJsonFile(manifestLocation);
});

export const loadJsonFileSync = vi.fn((manifestLocation) => {
  incrementCalled(syncRegistry, manifestLocation);

  return loadJsonFileSyncActual(manifestLocation);
});

(loadJsonFile as any).registry = asyncRegistry;
(loadJsonFileSync as any).registry = syncRegistry;

// keep test data isolated
afterEach(() => {
  asyncRegistry.clear();
  syncRegistry.clear();
});

// export { asyncRegistry as registry, mockLoadJsonFile as loadJsonFile, mockLoadJsonFileSync as loadJsonFileSync };
// export {
//   registry: asyncRegistry,
//   loadJsonFile: mockLoadJsonFile,
//   loadJsonFileSync: mockLoadJsonFileSync,
// };
// export {
//   // registry: asyncRegistry,
//   loadJsonFile: mockLoadJsonFile,
//   loadJsonFileSync: mockLoadJsonFileSync,
// }

// module.exports.loadJsonFile = mockLoadJsonFile;
// module.exports.registry = asyncRegistry;
// module.exports.loadJsonFileSync = mockLoadJsonFileSync;
// module.exports.loadJsonFileSync.registry = syncRegistry;

// mockLoadJsonFile.registry = asyncRegistry;
// mockLoadJsonFile.sync = mockLoadJsonFileSync;
// mockLoadJsonFile.sync.registry = syncRegistry;
// export default mockLoadJsonFile;
