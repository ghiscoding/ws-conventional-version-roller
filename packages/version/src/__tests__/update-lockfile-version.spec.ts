import { describe, expect, it, Mock, test, vi } from 'vitest';

vi.mock('load-json-file', async () => await vi.importActual('../lib/__mocks__/load-json-file'));
vi.mock('@lerna-lite/core', async () => {
  const { exec, execSync } = await vi.importActual<any>('@lerna-lite/core');
  return {
    ...(await vi.importActual<any>('@lerna-lite/core')),
    exec: vi.fn(exec),
    execSync: vi.fn(execSync),
  };
});

import npmlog from 'npmlog';
import { pathExistsSync, readJsonSync } from 'fs-extra/esm';
import { promises as fsPromises } from 'node:fs';
import { dirname as pathDirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { Package } from '@lerna-lite/core';

// mocked or stubbed modules
import { loadJsonFile } from 'load-json-file';

// helpers
const __filename = fileURLToPath(import.meta.url);
const __dirname = pathDirname(__filename);
import { exec, execSync, Project } from '@lerna-lite/core';
import { initFixtureFactory } from '@lerna-test/helpers';
const initFixture = initFixtureFactory(__dirname);

import {
  loadPackageLockFileWhenExists,
  updateClassicLockfileVersion,
  updateTempModernLockfileVersion,
  saveUpdatedLockJsonFile,
  runInstallLockFileOnly,
  validateFileExists,
} from '../lib/update-lockfile-version';

describe('npm classic lock file', () => {
  test('updateLockfileVersion with lockfile v1', async () => {
    const cwd = await initFixture('lockfile-leaf');
    const [pkg] = await Project.getPackages(cwd);

    pkg.version = '2.0.0';

    const returnedLockfilePath = await updateClassicLockfileVersion(pkg as unknown as Package);

    expect(returnedLockfilePath).toBe(join(pkg.location, 'package-lock.json'));
    expect(Array.from((loadJsonFile as any).registry.keys())).toStrictEqual(['/packages/package-1']);
    expect(readJsonSync(returnedLockfilePath as string)).toHaveProperty('version', '2.0.0');
  });

  test('updateClassicLockfileVersion with lockfile v2', async () => {
    const cwd = await initFixture('lockfile-leaf-v2');
    const [pkg] = await Project.getPackages(cwd);

    pkg.version = '2.0.0';
    pkg.dependencies['package-1'] = '^2.0.0';
    pkg.devDependencies['package-2'] = '3.0.0';

    const returnedLockfilePath = await updateClassicLockfileVersion(pkg as unknown as Package);

    const updatedLockfile = readJsonSync(returnedLockfilePath as string);
    expect(returnedLockfilePath).toBe(join(pkg.location, 'package-lock.json'));
    expect(Array.from((loadJsonFile as any).registry.keys())).toStrictEqual(['/packages/package-1']);
    expect(updatedLockfile).toHaveProperty('version', '2.0.0');
    expect(updatedLockfile.packages[''].dependencies['package-1']).toBe('^2.0.0');
    expect(updatedLockfile.packages[''].dependencies['tiny-tarball']).toBe('^1.0.0');
    expect(updatedLockfile.packages[''].devDependencies['package-2']).toBe('3.0.0');
  });

  test('updateClassicLockfileVersion without sibling lockfile', async () => {
    const cwd = await initFixture('lifecycle', false);
    const [pkg] = await Project.getPackages(cwd);

    pkg.version = '1.1.0';

    (loadJsonFile as any).mockImplementationOnce(() => Promise.reject(new Error('file not found')));

    const returnedLockfilePath = await updateClassicLockfileVersion(pkg as unknown as Package);

    expect(returnedLockfilePath).toBeUndefined();
    expect(pathExistsSync(join(pkg.location, 'package-lock.json'))).toBe(false);
  });
});

describe('npm modern lock file', () => {
  test('updateModernLockfileVersion v2 in project root', async () => {
    const mockVersion = '2.4.0';
    const cwd = await initFixture('lockfile-version2');
    const rootLockFilePath = join(cwd, 'package-lock.json');
    const packages = await Project.getPackages(cwd);

    const lockFileOutput = await loadPackageLockFileWhenExists(cwd);
    if (lockFileOutput!.json) {
      for (const pkg of packages) {
        pkg.version = mockVersion;
        await updateTempModernLockfileVersion(pkg as unknown as Package, lockFileOutput!.json);
      }
      await saveUpdatedLockJsonFile(lockFileOutput!.path, lockFileOutput!.json);
    }

    // expect(Array.from((loadJsonFile as any).registry.keys())).toStrictEqual(['/packages/package-1', '/packages/package-2', '/']);
    expect(readJsonSync(rootLockFilePath)).toMatchSnapshot();
  });
});

describe('validateFileExists() method', () => {
  it(`should return true when file exist`, async () => {
    const cwd = await initFixture('lockfile-version2');
    const exists = await validateFileExists(join(cwd, 'package-lock.json'));

    expect(exists).toBe(true);
  });

  it(`should return false when file does not exist`, async () => {
    const cwd = await initFixture('lockfile-version2');
    const exists = await validateFileExists(join(cwd, 'wrong-file.json'));

    expect(exists).toBe(false);
  });
});

describe('run install lockfile-only', () => {
  describe('npm client', () => {
    it(`should update project root lockfile by calling npm script "npm install --package-lock-only" when npm version is >= 8.5.0`, async () => {
      (execSync as any).mockReturnValueOnce('8.5.0');
      const cwd = await initFixture('lockfile-version2');

      const lockFileOutput = await runInstallLockFileOnly('npm', cwd, []);

      expect(execSync).toHaveBeenCalledWith('npm', ['--version']);
      expect(exec).toHaveBeenCalledWith('npm', ['install', '--package-lock-only'], { cwd });
      expect(lockFileOutput).toBe('package-lock.json');
    });

    it(`should display a log error when npm version is below 8.5.0 and not actually sync anything`, async () => {
      (execSync as any).mockReturnValueOnce('8.4.0');
      const cwd = await initFixture('lockfile-version2');
      const logSpy = vi.spyOn(npmlog, 'error');

      await runInstallLockFileOnly('npm', cwd, []);

      expect(execSync).toHaveBeenCalledWith('npm', ['--version']);
      expect(logSpy).toHaveBeenCalledWith(
        'lock',
        expect.stringContaining('your npm version is lower than 8.5.0 which is the minimum requirement to use `--sync-workspace-lock`')
      );
    });

    it(`should update project root lockfile by calling npm script "npm install --package-lock-only" with extra npm client arguments when provided`, async () => {
      (execSync as any).mockReturnValueOnce('8.5.0');
      const cwd = await initFixture('lockfile-version2');

      const lockFileOutput = await runInstallLockFileOnly('npm', cwd, ['--legacy-peer-deps']);

      expect(execSync).toHaveBeenCalledWith('npm', ['--version']);
      expect(exec).toHaveBeenCalledWith('npm', ['install', '--package-lock-only', '--legacy-peer-deps'], { cwd });
      expect(lockFileOutput).toBe('package-lock.json');
    });

    it(`should update project root lockfile by calling npm script "npm install --legacy-peer-deps,--force" with multiple npm client arguments provided as CSV`, async () => {
      (execSync as any).mockReturnValueOnce('8.5.0');
      const cwd = await initFixture('lockfile-version2');

      const lockFileOutput = await runInstallLockFileOnly('npm', cwd, ['--legacy-peer-deps,--force']);

      expect(execSync).toHaveBeenCalledWith('npm', ['--version']);
      expect(exec).toHaveBeenCalledWith('npm', ['install', '--package-lock-only', '--legacy-peer-deps', '--force'], {
        cwd,
      });
      expect(lockFileOutput).toBe('package-lock.json');
    });
  });

  describe('pnpm client', () => {
    it('should log an error when lockfile is not located under project root', async () => {
      const logSpy = vi.spyOn(npmlog, 'error');
      const cwd = await initFixture('lockfile-version2');

      const lockFileOutput = await runInstallLockFileOnly('pnpm', cwd, []);

      expect(logSpy).toHaveBeenCalledWith(
        'lock',
        expect.stringContaining(`we could not sync neither locate "pnpm-lock.yaml" by using "pnpm" client at location ${cwd}`)
      );
      expect(lockFileOutput).toBe(undefined);
    });

    it(`should update project root lockfile by calling client script "pnpm install --package-lock-only"`, async () => {
      vi.spyOn(fsPromises, 'access').mockResolvedValue(true as any);
      (exec as Mock).mockImplementationOnce(() => true);
      const cwd = await initFixture('lockfile-version2');

      const lockFileOutput = await runInstallLockFileOnly('pnpm', cwd, []);

      expect(exec).toHaveBeenCalledWith('pnpm', ['install', '--lockfile-only', '--ignore-scripts'], { cwd });
      expect(lockFileOutput).toBe('pnpm-lock.yaml');
    });

    it(`should update project root lockfile by calling client script "pnpm install --package-lock-only" with extra npm client arguments when provided`, async () => {
      vi.spyOn(fsPromises, 'access').mockResolvedValue(true as any);
      (exec as Mock).mockImplementationOnce(() => true);
      const cwd = await initFixture('lockfile-version2');

      const lockFileOutput = await runInstallLockFileOnly('pnpm', cwd, ['--frozen-lockfile']);

      expect(exec).toHaveBeenCalled();
      expect(exec).toHaveBeenCalledWith('pnpm', ['install', '--lockfile-only', '--ignore-scripts', '--frozen-lockfile'], { cwd });
      expect(lockFileOutput).toBe('pnpm-lock.yaml');
    });
  });

  describe('yarn client', () => {
    it(`should NOT update project root lockfile when yarn version is 1.0.0 and is below 2.0.0`, async () => {
      (execSync as any).mockReturnValueOnce('1.0.0');
      vi.spyOn(fsPromises, 'access').mockResolvedValue(true as any);
      (exec as Mock).mockImplementationOnce(() => true);
      const cwd = await initFixture('lockfile-version2');

      await runInstallLockFileOnly('yarn', cwd, []);

      expect(execSync).toHaveBeenCalledWith('yarn', ['--version']);
      expect(exec).not.toHaveBeenCalled();
    });

    it(`should update project root lockfile by calling client script "yarn install --package-lock-only"`, async () => {
      (execSync as any).mockReturnValueOnce('3.0.0');
      vi.spyOn(fsPromises, 'access').mockResolvedValue(true as any);
      (exec as Mock).mockImplementationOnce(() => true);
      const cwd = await initFixture('lockfile-version2');

      const lockFileOutput = await runInstallLockFileOnly('yarn', cwd, []);

      expect(execSync).toHaveBeenCalledWith('yarn', ['--version']);
      expect(exec).toHaveBeenCalledWith('yarn', ['install', '--mode', 'update-lockfile'], { cwd });
      expect(lockFileOutput).toBe('yarn.lock');
    });

    it(`should update project root lockfile by calling client script "yarn install --package-lock-only" with extra npm client arguments when provided`, async () => {
      (execSync as any).mockReturnValueOnce('4.0.0');
      vi.spyOn(fsPromises, 'access').mockResolvedValue(true as any);
      (exec as Mock).mockImplementationOnce(() => true);
      const cwd = await initFixture('lockfile-version2');

      const lockFileOutput = await runInstallLockFileOnly('yarn', cwd, ['--check-files']);

      expect(execSync).toHaveBeenCalledWith('yarn', ['--version']);
      expect(exec).toHaveBeenCalledWith('yarn', ['install', '--mode', 'update-lockfile', '--check-files'], { cwd });
      expect(lockFileOutput).toBe('yarn.lock');
    });
  });
});
