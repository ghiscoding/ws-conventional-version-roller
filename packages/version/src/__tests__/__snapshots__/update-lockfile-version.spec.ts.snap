// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`npm modern lock file > updateModernLockfileVersion v2 in project root 1`] = `
{
  'dependencies': {
    '@my-workspace/package-1': {
      'requires': {
        'tiny-tarball': '^1.0.0',
      },
      'version': 'file:packages/package-1',
    },
    '@my-workspace/package-2': {
      'requires': {
        '@my-workspace/package-1': '^2.4.0',
      },
      'version': 'file:packages/package-2',
    },
  },
  'lockfileVersion': 2,
  'name': 'my-workspace',
  'packages': {
    '': {
      'license': 'MIT',
      'name': 'my-workspace',
      'workspaces': [
        './packages/package-1',
        './packages/package-2',
      ],
    },
    'node_modules/package-1': {
      'link': true,
      'resolved': 'packages/package-1',
    },
    'node_modules/package-2': {
      'link': true,
      'resolved': 'packages/package-2',
    },
    'packages/package-1': {
      'license': 'MIT',
      'name': '@my-workspace/package-1',
      'tiny-tarball': {
        'integrity': 'sha1-u/EC1a5zr+LFUyleD7AiMCFvZbE=',
        'resolved': 'https://registry.npmjs.org/tiny-tarball/-/tiny-tarball-1.0.0.tgz',
        'version': '1.0.0',
      },
      'version': '2.4.0',
    },
    'packages/package-2': {
      'dependencies': {
        '@my-workspace/package-1': '^2.4.0',
      },
      'license': 'MIT',
      'name': '@my-workspace/package-2',
      'version': '2.4.0',
    },
  },
  'requires': true,
}
`;
